File: cpp/src/local_planning/velocity/agent_sim.cpp
#include "navigation/local_planning/velocity/agent_sim.hpp"

#include <algorithm>
#include <cassert>
#include <cmath>

namespace navigation::local_planning {

namespace {

double pairwise_min_clearance(const std::vector<core::RobotState>& states,
                              const std::vector<AgentSpec>& specs) {
  double best = std::numeric_limits<double>::infinity();
  size_t n = states.size();
  for (size_t i = 0; i < n; ++i) {
    for (size_t j = i + 1; j < n; ++j) {
      double gap = std::hypot(states[i].pose.x - states[j].pose.x,
                              states[i].pose.y - states[j].pose.y) -
                   specs[i].radius - specs[j].radius;
      if (gap < best) best = gap;
    }
  }
  return best;
}

}  // namespace

std::vector<AgentResult> simulate_agents(const std::vector<VelocityObstaclePlanner*>& planners,
                                         const std::vector<AgentSpec>& specs,
                                         core::ObstacleQuery& space, const SimConfig& config,
                                         core::TraceRecorder* recorder) {
  const size_t n = specs.size();
  const core::Footprint footprint{config.footprint_radius};
  std::vector<core::RobotState> states;
  states.reserve(n);
  for (const AgentSpec& spec : specs) states.push_back(spec.start);
  std::vector<core::Point> world_vel(n, core::Point{0.0, 0.0});
  std::vector<std::vector<core::Pose>> trajectories(n);
  for (size_t k = 0; k < n; ++k) trajectories[k].push_back(specs[k].start.pose);
  std::vector<bool> reached(n, false);
  std::vector<size_t> planner_indices;
  for (size_t k = 0; k < n; ++k) {
    if (!specs[k].scripted_velocity.has_value()) planner_indices.push_back(k);
  }

  double min_pair_clearance = pairwise_min_clearance(states, specs);
  SimStatus terminal = SimStatus::TIMEOUT;
  int steps = config.max_steps;

  for (int step = 1; step <= config.max_steps; ++step) {
    // 1) snapshot every body before this tick's motion.
    std::vector<DynamicObstacle> snapshot;
    snapshot.reserve(n);
    for (size_t k = 0; k < n; ++k) {
      snapshot.push_back(DynamicObstacle{
          core::Point{states[k].pose.x, states[k].pose.y}, world_vel[k], specs[k].radius});
    }
    // 2) compute every command against that fixed snapshot, index order.
    std::vector<std::optional<core::VelocityCommand>> commands(n);
    for (size_t k : planner_indices) {
      std::vector<DynamicObstacle> neighbors;
      neighbors.reserve(n > 0 ? n - 1 : 0);
      for (size_t j = 0; j < n; ++j) {
        if (j != k) neighbors.push_back(snapshot[j]);
      }
      core::LocalTask task{specs[k].goal, {}};
      VelocityObstaclePlanner* planner = planners[k];
      assert(planner != nullptr && "agent has no planner but is not scripted");
      core::TraceRecorder* rec = (k == 0) ? recorder : nullptr;
      commands[k] =
          planner->command_with_neighbors(space, states[k], task, neighbors, config.control_dt, rec);
    }
    // 3) integrate every body only after every command is known.
    std::vector<core::RobotState> new_states = states;
    for (size_t k = 0; k < n; ++k) {
      if (specs[k].scripted_velocity.has_value()) {
        core::Point v = *specs[k].scripted_velocity;
        double x = states[k].pose.x, y = states[k].pose.y;
        core::Pose new_pose{x + v.x * config.control_dt, y + v.y * config.control_dt,
                            std::atan2(v.y, v.x)};
        world_vel[k] = v;
        new_states[k] = core::RobotState{new_pose, std::hypot(v.x, v.y), 0.0};
      } else {
        const core::VelocityCommand& cmd = *commands[k];
        core::Pose new_pose = integrate_unicycle(states[k].pose, cmd, config.control_dt);
        world_vel[k] = core::Point{cmd.v * std::cos(new_pose.theta), cmd.v * std::sin(new_pose.theta)};
        new_states[k] = core::RobotState{new_pose, cmd.v, cmd.omega};
      }
    }
    states = new_states;
    // 4) trace + trajectory bookkeeping for every body.
    for (size_t k = 0; k < n; ++k) {
      trajectories[k].push_back(states[k].pose);
      if (recorder != nullptr) {
        recorder->robot_moved(
            states[k].pose,
            core::TraceRecorder::EventData{{"v", states[k].v}, {"omega", states[k].omega}},
            static_cast<int>(k));
      }
    }
    // 5) termination judgement.
    double tick_clearance = pairwise_min_clearance(states, specs);
    min_pair_clearance = std::min(min_pair_clearance, tick_clearance);
    bool collided = tick_clearance < 0.0;
    if (!collided) {
      for (size_t k = 0; k < n; ++k) {
        if (space.is_collision(footprint, states[k].pose)) {
          collided = true;
          break;
        }
      }
    }
    if (collided) {
      terminal = SimStatus::COLLISION;
      steps = step;
      break;
    }
    for (size_t k : planner_indices) {
      if (!reached[k] &&
          std::hypot(states[k].pose.x - specs[k].goal.x, states[k].pose.y - specs[k].goal.y) <=
              config.goal_tolerance) {
        reached[k] = true;
      }
    }
    bool all_reached =
        std::all_of(planner_indices.begin(), planner_indices.end(), [&](size_t k) { return reached[k]; });
    if (all_reached) {
      terminal = SimStatus::REACHED;
      steps = step;
      break;
    }
    std::vector<size_t> still_active;
    for (size_t k : planner_indices) {
      if (!reached[k]) still_active.push_back(k);
    }
    if (!still_active.empty() && step >= config.stall_window) {
      bool all_stalled = true;
      for (size_t k : still_active) {
        const core::Pose& past = trajectories[k][static_cast<size_t>(step - config.stall_window)];
        double moved = std::hypot(states[k].pose.x - past.x, states[k].pose.y - past.y);
        if (moved >= config.stall_distance) {
          all_stalled = false;
          break;
        }
      }
      if (all_stalled) {
        terminal = SimStatus::STALLED;
        steps = step;
        break;
      }
    }
    steps = step;
  }

  std::vector<AgentResult> results;
  results.reserve(n);
  for (size_t k = 0; k < n; ++k) {
    results.push_back(AgentResult{reached[k] ? SimStatus::REACHED : terminal, steps,
                                  trajectories[k], min_pair_clearance});
  }
  return results;
}

}  // namespace navigation::local_planning

File: python/navigation/local_planning/velocity/agent_sim.py
"""Multi-agent closed-loop harness for the velocity-obstacle family.

N bodies share one map; each is either driven by its own VelocityObstaclePlanner
or moves at a fixed scripted velocity (a non-cooperative mover for single-planner
demos). Every tick is read-all-then-write: every body's command is computed
against the SAME pre-tick snapshot of every other body, then all bodies
integrate together -- order-independent and reproducible, which reciprocal
avoidance (van den Berg et al. 2008) depends on (a sequential update would make
agent 0 react to agent 1's *old* position while agent 1 reacts to agent 0's
*new* one, breaking the reciprocity the algorithm assumes).

Reuses `local_planning/simulation.py`'s `integrate_unicycle`/`SimConfig`/
`SimStatus` (family -> category-shared root, the same direction `_geometry.py`
is imported from) rather than re-deriving single-body integration or status
enums here.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field

from navigation.core.capabilities import ObstacleQuery
from navigation.core.trace import TraceRecorder
from navigation.core.types import Footprint, LocalTask, Point, Pose, RobotState, VelocityCommand
from navigation.local_planning.simulation import SimConfig, SimStatus, integrate_unicycle

from ._velocity_obstacle import DynamicObstacle, VelocityObstaclePlanner


@dataclass(frozen=True)
class AgentSpec:
    start: RobotState
    goal: Pose
    radius: float
    # None -> driven by the matching planner; a value -> a non-cooperative
    # constant-velocity mover (the VO/RVO/ORCA demos' scripted crossing traffic).
    scripted_velocity: Point | None = None


@dataclass
class AgentResult:
    status: SimStatus
    steps: int
    trajectory: list[Pose] = field(default_factory=list)
    min_pair_clearance: float = math.inf


def _xy_dist(a: tuple[float, float] | Pose, b: tuple[float, float] | Pose) -> float:
    return math.hypot(a[0] - b[0], a[1] - b[1])


def _pairwise_min_clearance(states: list[RobotState], specs: list[AgentSpec]) -> float:
    best = math.inf
    n = len(states)
    for i in range(n):
        for j in range(i + 1, n):
            gap = _xy_dist(states[i].pose, states[j].pose) - specs[i].radius - specs[j].radius
            if gap < best:
                best = gap
    return best


def simulate_agents(
    planners: list[VelocityObstaclePlanner | None],
    specs: list[AgentSpec],
    space: ObstacleQuery,
    config: SimConfig,
    recorder: TraceRecorder | None = None,
) -> list[AgentResult]:
    """Run every agent in closed loop until every planner-driven agent reaches
    its goal, a collision (pairwise or static) occurs, the group stalls, or
    the step budget runs out. `planners[k] is None` iff `specs[k]` is scripted.

    Reached agents are not special-cased out of planning/integration: once an
    agent is near its own goal, `preferred_velocity` alone drives it toward
    (0, 0), so it settles in place naturally and keeps contributing an
    (approximately stationary) DynamicObstacle snapshot to its neighbors.
    """
    n = len(specs)
    footprint = Footprint(config.footprint_radius)
    states = [spec.start for spec in specs]
    world_vel: list[Point] = [(0.0, 0.0) for _ in range(n)]
    trajectories: list[list[Pose]] = [[spec.start.pose] for spec in specs]
    reached: list[bool] = [False] * n
    planner_indices = [k for k in range(n) if specs[k].scripted_velocity is None]

    min_pair_clearance = _pairwise_min_clearance(states, specs)
    terminal = SimStatus.TIMEOUT
    steps = config.max_steps

    for step in range(1, config.max_steps + 1):
        # 1) snapshot every body before this tick's motion.
        snapshot = [
            DynamicObstacle(
                position=(states[k].pose[0], states[k].pose[1]),
                velocity=world_vel[k],
                radius=specs[k].radius,
            )
            for k in range(n)
        ]
        # 2) compute every command against that fixed snapshot, index order.
        commands: list[VelocityCommand | None] = [None] * n
        for k in planner_indices:
            neighbors = tuple(snapshot[j] for j in range(n) if j != k)
            task = LocalTask(goal=specs[k].goal)
            planner = planners[k]
            assert planner is not None, f"agent {k} has no planner but is not scripted"
            rec = recorder if k == 0 else None
            commands[k] = planner.command_with_neighbors(
                space, states[k], task, neighbors, config.control_dt, rec
            )
        # 3) integrate every body only after every command is known.
        new_states = list(states)
        for k in range(n):
            scripted = specs[k].scripted_velocity
            if scripted is not None:
                vx, vy = scripted
                x, y, _theta = states[k].pose
                new_pose = (
                    x + vx * config.control_dt,
                    y + vy * config.control_dt,
                    math.atan2(vy, vx),
                )
                world_vel[k] = (vx, vy)
                new_states[k] = RobotState(new_pose, math.hypot(vx, vy), 0.0)
            else:
                cmd = commands[k]
                assert cmd is not None
                new_pose = integrate_unicycle(states[k].pose, cmd, config.control_dt)
                world_vel[k] = (cmd.v * math.cos(new_pose[2]), cmd.v * math.sin(new_pose[2]))
                new_states[k] = RobotState(new_pose, cmd.v, cmd.omega)
        states = new_states
        # 4) trace + trajectory bookkeeping for every body.
        for k in range(n):
            trajectories[k].append(states[k].pose)
            if recorder is not None:
                recorder.robot_moved(
                    states[k].pose, data={"v": states[k].v, "omega": states[k].omega}, agent=k
                )
        # 5) termination judgement.
        tick_clearance = _pairwise_min_clearance(states, specs)
        min_pair_clearance = min(min_pair_clearance, tick_clearance)
        collided = tick_clearance < 0.0 or any(
            space.is_collision(footprint, states[k].pose) for k in range(n)
        )
        if collided:
            terminal, steps = SimStatus.COLLISION, step
            break
        for k in planner_indices:
            if not reached[k] and _xy_dist(states[k].pose, specs[k].goal) <= config.goal_tolerance:
                reached[k] = True
        if all(reached[k] for k in planner_indices):
            terminal, steps = SimStatus.REACHED, step
            break
        still_active = [k for k in planner_indices if not reached[k]]
        if (
            still_active
            and step >= config.stall_window
            and all(
                _xy_dist(states[k].pose, trajectories[k][step - config.stall_window])
                < config.stall_distance
                for k in still_active
            )
        ):
            terminal, steps = SimStatus.STALLED, step
            break
        steps = step

    return [
        AgentResult(
            status=SimStatus.REACHED if reached[k] else terminal,
            steps=steps,
            trajectory=list(trajectories[k]),
            min_pair_clearance=min_pair_clearance,
        )
        for k in range(n)
    ]
